### Importing Libraries

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

In [3]:
df=pd.read_csv("ball_by_ball_it20.csv")

In [4]:
df

Unnamed: 0  Match ID        Date            Venue    Bat First  \
0                0   1339605  2023-03-26  SuperSport Park  West Indies   
1                1   1339605  2023-03-26  SuperSport Park  West Indies   
2                2   1339605  2023-03-26  SuperSport Park  West Indies   
3                3   1339605  2023-03-26  SuperSport Park  West Indies   
4                4   1339605  2023-03-26  SuperSport Park  West Indies   
...            ...       ...         ...              ...          ...   
425114      425114   1187667  2019-11-05      Saxton Oval  New Zealand   
425115      425115   1187667  2019-11-05      Saxton Oval  New Zealand   
425116      425116   1187667  2019-11-05      Saxton Oval  New Zealand   
425117      425117   1187667  2019-11-05      Saxton Oval  New Zealand   
425118      425118   1187667  2019-11-05      Saxton Oval  New Zealand   

          Bat Second  Innings  Over  Ball     Batter  ...        Winner  \
0       South Africa        1     1     1    BA King  ...  South Africa   
1       South Africa        1     1     2  KR Mayers  ...  South Africa   
2       South Africa        1     1     3    BA King  ...  South Africa   
3       South Africa        1     1     4  J Charles  ...  South Africa   
4       South Africa        1     1     5  J Charles  ...  South Africa   
...              ...      ...   ...   ...        ...  ...           ...   
425114       England        2    20     2  TK Curran  ...   New Zealand   
425115       England        2    20     3  TK Curran  ...   New Zealand   
425116       England        2    20     4  S Mahmood  ...   New Zealand   
425117       England        2    20     5  S Mahmood  ...   New Zealand   
425118       England        2    20     6  TK Curran  ...   New Zealand   

       Chased Successfully  Total Batter Runs  Total Non Striker Runs  \
0                        1                  1                       0   
1                        1                  1                       1   
2                        1                  0                       1   
3                        1                  0                       1   
4                        1                  4                       1   
...                    ...                ...                     ...   
425114                   0                 11                       2   
425115                   0                 12                       2   
425116                   0                  2                      12   
425117                   0                  3                      12   
425118                   0                 14                       3   

        Batter Balls Faced  Non Striker Balls Faced Player Out Runs  \
0                        1                        0             NaN   
1                        1                        1             NaN   
2                        0                        1             1.0   
3                        1                        1             NaN   
4                        2                        1             NaN   
...                    ...                      ...             ...   
425114                   9                        2             NaN   
425115                  10                        2             NaN   
425116                   3                       10             NaN   
425117                   4                       10             NaN   
425118                  11                        4             NaN   

        Player Out Balls Faced Bowler Runs Conceded Valid Ball  
0                          NaN                    1          1  
1                          NaN                    1          1  
2                          2.0                    0          1  
3                          NaN                    0          1  
4                          NaN                    4          1  
...                        ...                  ...        ...  
425114                     NaN   

In [5]:
df.shape

(425119, 35)

### Including only those 20 teams and their matches which are playing in T20 world cup 2024

In [6]:
sorted(df['Bat First'].unique())

['Afghanistan',
 'Argentina',
 'Australia',
 'Austria',
 'Bahamas',
 'Bahrain',
 'Bangladesh',
 'Belgium',
 'Belize',
 'Bermuda',
 'Bhutan',
 'Botswana',
 'Bulgaria',
 'Cameroon',
 'Canada',
 'Cayman Islands',
 'China',
 'Cook Islands',
 'Croatia',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'England',
 'Estonia',
 'Eswatini',
 'Fiji',
 'Finland',
 'France',
 'Gambia',
 'Germany',
 'Ghana',
 'Gibraltar',
 'Greece',
 'Guernsey',
 'Hong Kong',
 'Hungary',
 'India',
 'Indonesia',
 'Iran',
 'Ireland',
 'Isle of Man',
 'Israel',
 'Italy',
 'Japan',
 'Jersey',
 'Kenya',
 'Kuwait',
 'Lesotho',
 'Luxembourg',
 'Malawi',
 'Malaysia',
 'Maldives',
 'Mali',
 'Malta',
 'Mozambique',
 'Myanmar',
 'Namibia',
 'Nepal',
 'Netherlands',
 'New Zealand',
 'Nigeria',
 'Norway',
 'Oman',
 'Pakistan',
 'Panama',
 'Papua New Guinea',
 'Philippines',
 'Portugal',
 'Qatar',
 'Romania',
 'Rwanda',
 'Samoa',
 'Saudi Arabia',
 'Scotland',
 'Serbia',
 'Seychelles',
 'Sierra Leone',
 'Singapore',
 'Slovenia',
 'Sout

In [7]:
teams=['Afghanistan','Australia','Bangladesh','Canada','England','India','Ireland','Namibia','Nepal','Netherlands',
       'New Zealand','Oman','Pakistan','Papua New Guinea','Scotland','South Africa','Sri Lanka','Uganda',
      'United States of America','West Indies']

In [8]:
df=df[df['Bat First'].isin(teams)]
df=df[df['Bat Second'].isin(teams)]

In [9]:
df.shape

(218212, 35)

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 218212 entries, 0 to 425118
Data columns (total 35 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Unnamed: 0               218212 non-null  int64  
 1   Match ID                 218212 non-null  int64  
 2   Date                     218212 non-null  object 
 3   Venue                    218212 non-null  object 
 4   Bat First                218212 non-null  object 
 5   Bat Second               218212 non-null  object 
 6   Innings                  218212 non-null  int64  
 7   Over                     218212 non-null  int64  
 8   Ball                     218212 non-null  int64  
 9   Batter                   218212 non-null  object 
 10  Non Striker              218212 non-null  object 
 11  Bowler                   218212 non-null  object 
 12  Batter Runs              218212 non-null  int64  
 13  Extra Runs               218212 non-null  int64  
 14  Runs From

In [11]:
df=df[['Match ID','Bat First','Bat Second','Over','Ball','Runs From Ball','Wicket','Target Score']]

In [12]:
df

Match ID    Bat First    Bat Second  Over  Ball  Runs From Ball  \
0        1339605  West Indies  South Africa     1     1               1   
1        1339605  West Indies  South Africa     1     2               1   
2        1339605  West Indies  South Africa     1     3               0   
3        1339605  West Indies  South Africa     1     4               0   
4        1339605  West Indies  South Africa     1     5               4   
...          ...          ...           ...   ...   ...             ...   
425114   1187667  New Zealand       England    20     2               0   
425115   1187667  New Zealand       England    20     3               1   
425116   1187667  New Zealand       England    20     4               0   
425117   1187667  New Zealand       England    20     5               1   
425118   1187667  New Zealand       England    20     6               2   

        Wicket  Target Score  
0            0           259  
1            0           259  
2            1           259  
3            0           259  
4            0           259  
...        ...           ...  
425114       0           181  
425115       0           181  
425116       0           181  
425117       0           181  
425118       0           181  

[218212 rows x 8 columns]

In [13]:
df.isnull().sum()

Match ID          0
Bat First         0
Bat Second        0
Over              0
Ball              0
Runs From Ball    0
Wicket            0
Target Score      0
dtype: int64

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 218212 entries, 0 to 425118
Data columns (total 8 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   Match ID        218212 non-null  int64 
 1   Bat First       218212 non-null  object
 2   Bat Second      218212 non-null  object
 3   Over            218212 non-null  int64 
 4   Ball            218212 non-null  int64 
 5   Runs From Ball  218212 non-null  int64 
 6   Wicket          218212 non-null  int64 
 7   Target Score    218212 non-null  int64 
dtypes: int64(6), object(2)
memory usage: 15.0+ MB


### calculating the current score by using cumsum() function

In [15]:
df['Current_score']=df.groupby('Match ID')['Runs From Ball'].cumsum()
df=df[df['Current_score']<df['Target Score']]

In [16]:
df['Wicket'].value_counts()

Wicket
0    107596
1      6047
Name: count, dtype: int64

### creating wickets_left column 

In [17]:
df['wickets_left']=10-df.groupby('Match ID')['Wicket'].cumsum()

### balls_left column

In [18]:
df['Over']=df['Over']-1
df['balls_bowled']= (df['Over']*6) + df['Ball'] 

### Cleaning up unncessary current_score values

In [19]:
df_0=df[(df['Over']==0) & (df['Current_score']>36)]
df_1=df[(df['Over']==1) & (df['Current_score']>72)]
df_2=df[(df['Over']==2) & (df['Current_score']>100)]

In [20]:
dfs_index_drop= df_0.index.union(df_1.index).union(df_2.index)
df=df.drop(dfs_index_drop)

In [21]:
df.describe()

Match ID           Over           Ball  Runs From Ball  \
count  1.125250e+05  112525.000000  112525.000000   112525.000000   
mean   9.585564e+05       9.336521       3.493082        1.301337   
std    3.570615e+05       5.753345       1.707983        1.580611   
min    2.110280e+05       0.000000       0.000000        0.000000   
25%    5.730270e+05       4.000000       2.000000        0.000000   
50%    1.133820e+06       9.000000       3.000000        1.000000   
75%    1.272095e+06      14.000000       5.000000        1.000000   
max    1.384637e+06      19.000000       7.000000        7.000000   

              Wicket   Target Score  Current_score   wickets_left  \
count  112525.000000  112525.000000  112525.000000  112525.000000   
mean        0.053153     160.267763      75.396161       7.392286   
std         0.224339      34.964814      49.797768       2.140081   
min         0.000000      40.000000       0.000000       0.000000   
25%         0.000000     138.000000      34.000000       6.000000   
50%         0.000000     160.000000      70.000000       8.000000   
75%         0.000000     184.000000     111.000000       9.000000   
max         1.000000     279.000000     278.000000      10.000000   

        balls_bowled  
count  112525.000000  
mean       59.512206  
std        34.561250  
min         1.000000  
25%        29.000000  
50%        59.000000  
75%        89.000000  
max       121.000000

### creating balls_left column

In [22]:
df['balls_left']=120-df['balls_bowled']

In [23]:
df['balls_bowled'].unique()

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103, 104,
       105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
       118, 119, 120, 121], dtype=int64)

In [24]:
df['balls_left']= df['balls_left'].apply(lambda x:0 if x<0 else x)

In [25]:
df['balls_left'].unique()

array([119, 118, 117, 116, 115, 114, 113, 112, 111, 110, 109, 108, 107,
       106, 105, 104, 103, 102, 101, 100,  99,  98,  97,  96,  95,  94,
        93,  92,  91,  90,  89,  88,  87,  86,  85,  84,  83,  82,  81,
        80,  79,  78,  77,  76,  75,  74,  73,  72,  71,  70,  69,  68,
        67,  66,  65,  64,  63,  62,  61,  60,  59,  58,  57,  56,  55,
        54,  53,  52,  51,  50,  49,  48,  47,  46,  45,  44,  43,  42,
        41,  40,  39,  38,  37,  36,  35,  34,  33,  32,  31,  30,  29,
        28,  27,  26,  25,  24,  23,  22,  21,  20,  19,  18,  17,  16,
        15,  14,  13,  12,  11,  10,   9,   8,   7,   6,   5,   4,   3,
         2,   1,   0], dtype=int64)

### calculating current run rate (crr)

In [26]:
df['crr']= round((df['Current_score']*6)/df['balls_bowled'],3)
df

Match ID    Bat First    Bat Second  Over  Ball  Runs From Ball  \
0        1339605  West Indies  South Africa     0     1               1   
1        1339605  West Indies  South Africa     0     2               1   
2        1339605  West Indies  South Africa     0     3               0   
3        1339605  West Indies  South Africa     0     4               0   
4        1339605  West Indies  South Africa     0     5               4   
...          ...          ...           ...   ...   ...             ...   
424989   1187667  New Zealand       England    19     2               1   
424990   1187667  New Zealand       England    19     3               2   
424991   1187667  New Zealand       England    19     4               4   
424992   1187667  New Zealand       England    19     5               1   
424993   1187667  New Zealand       England    19     6               0   

        Wicket  Target Score  Current_score  wickets_left  balls_bowled  \
0            0           259              1            10             1   
1            0           259              2            10             2   
2            1           259              2             9             3   
3            0           259              2             9             4   
4            0           259              6             9             5   
...        ...           ...            ...           ...           ...   
424989       0           181            173             4           116   
424990       0           181            175             4           117   
424991       0           181            179             4           118   
424992       0           181            180             4           119   
424993       1           181            180             3           120   

        balls_left    crr  
0              119  6.000  
1              118  6.000  
2              117  4.000  
3              116  3.000  
4              115  7.200  
...            ...    ...  
424989           4  8.948  
424990           3  8.974  
424991           2  9.102  
424992           1  9.076  
424993           0  9.000  

[112525 rows x 13 columns]

### calculating runs in last five overs

In [27]:
grps= df.groupby('Match ID')

ids= df['Match ID'].unique()
runs_five= []
for id in ids:
    runs_five.extend(grps.get_group(id).rolling(window=30)['Runs From Ball'].sum().values.tolist())

In [28]:
len(runs_five)

112525

In [29]:
df['runs_last_five']=runs_five

In [30]:
df

Match ID    Bat First    Bat Second  Over  Ball  Runs From Ball  \
0        1339605  West Indies  South Africa     0     1               1   
1        1339605  West Indies  South Africa     0     2               1   
2        1339605  West Indies  South Africa     0     3               0   
3        1339605  West Indies  South Africa     0     4               0   
4        1339605  West Indies  South Africa     0     5               4   
...          ...          ...           ...   ...   ...             ...   
424989   1187667  New Zealand       England    19     2               1   
424990   1187667  New Zealand       England    19     3               2   
424991   1187667  New Zealand       England    19     4               4   
424992   1187667  New Zealand       England    19     5               1   
424993   1187667  New Zealand       England    19     6               0   

        Wicket  Target Score  Current_score  wickets_left  balls_bowled  \
0            0           259              1            10             1   
1            0           259              2            10             2   
2            1           259              2             9             3   
3            0           259              2             9             4   
4            0           259              6             9             5   
...        ...           ...            ...           ...           ...   
424989       0           181            173             4           116   
424990       0           181            175             4           117   
424991       0           181            179             4           118   
424992       0           181            180             4           119   
424993       1           181            180             3           120   

        balls_left    crr  runs_last_five  
0              119  6.000             NaN  
1              118  6.000             NaN  
2              117  4.000             NaN  
3              116  3.000             NaN  
4              115  7.200             NaN  
...            ...    ...             ...  
424989           4  8.948            37.0  
424990           3  8.974            38.0  
424991           2  9.102            41.0  
424992           1  9.076            41.0  
424993           0  9.000            40.0  

[112525 rows x 14 columns]

### creating final df

In [31]:
df_final= df[['Bat First','Bat Second','Current_score','balls_left','wickets_left','crr','runs_last_five','Target Score']]

### Dropping null values

In [32]:
df_final = df_final.rename(columns={'Bat First': 'Bat_first', 'Bat Second': 'Bat_second'})
df_final.isnull().sum()

Bat_first             0
Bat_second            0
Current_score         0
balls_left            0
wickets_left          0
crr                   0
runs_last_five    27051
Target Score          0
dtype: int64

In [33]:
df_final.dropna(inplace=True)

In [34]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
Index: 85474 entries, 29 to 424993
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Bat_first       85474 non-null  object 
 1   Bat_second      85474 non-null  object 
 2   Current_score   85474 non-null  int64  
 3   balls_left      85474 non-null  int64  
 4   wickets_left    85474 non-null  int64  
 5   crr             85474 non-null  float64
 6   runs_last_five  85474 non-null  float64
 7   Target Score    85474 non-null  int64  
dtypes: float64(2), int64(4), object(2)
memory usage: 5.9+ MB


In [35]:
df_final.describe()

Current_score    balls_left  wickets_left           crr  \
count   85474.000000  85474.000000  85474.000000  85474.000000   
mean       93.993986     46.204109      6.739125      7.583791   
std        42.175378     26.516064      2.015950      1.772337   
min         8.000000      0.000000      0.000000      1.600000   
25%        60.000000     23.000000      5.000000      6.364000   
50%        89.000000     46.000000      7.000000      7.477000   
75%       123.000000     69.000000      8.000000      8.673000   
max       278.000000     98.000000     10.000000     16.000000   

       runs_last_five  Target Score  
count    85474.000000  85474.000000  
mean        38.607237    160.969394  
std         11.880169     34.353680  
min          8.000000     40.000000  
25%         30.000000    138.000000  
50%         38.000000    161.000000  
75%         46.000000    184.000000  
max         92.000000    279.000000

In [36]:
df_final= df_final.sample(df_final.shape[0])

In [37]:
df_final

Bat_first   Bat_second  Current_score  balls_left  wickets_left  \
409710     Australia      England            138           6             5   
120307      Scotland   Bangladesh             97          23             4   
89638       Pakistan    Australia            104          35             8   
316318     Sri Lanka     Pakistan             53          74             6   
267632         India    Australia             45          87             9   
...              ...          ...            ...         ...           ...   
409444         Nepal  Netherlands            143          34             8   
403300   New Zealand     Pakistan            175          12             5   
367553  South Africa     Pakistan            129          42             7   
377301     Sri Lanka    Australia            105          36             7   
110796   New Zealand    Australia             33          85             7   

          crr  runs_last_five  Target Score  
409710  7.263            52.0           148  
120307  6.000            44.0           141  
89638   7.341            51.0           150  
316318  6.913            32.0           171  
267632  8.182            43.0           141  
...       ...             ...           ...  
409444  9.977            56.0           239  
403300  9.722            54.0           197  
367553  9.923            62.0           189  
377301  7.500            35.0           199  
110796  5.657            27.0           185  

[85474 rows x 8 columns]

### train_test_split

In [38]:
X= df_final.drop(columns=['Target Score'])
y= df_final['Target Score']

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test= train_test_split(X,y,test_size=0.2,random_state=42)

In [39]:
X_test.shape

(17095, 7)

### model making 

In [40]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

from sklearn.model_selection import GridSearchCV,cross_val_score
from sklearn.metrics import r2_score,mean_absolute_error

In [41]:
trf= ColumnTransformer([
    ('trf',OneHotEncoder(sparse=False, drop='first'),['Bat_first','Bat_second'])
], remainder= 'passthrough')

### Random Forest

In [42]:
from sklearn.ensemble import RandomForestRegressor
param_grid= {
    'n_estimators': [25, 50, 100, 150], 
    'max_features': [0.5,0.75,1.0], 
    'max_depth': [3, 6, 9,], 
    'max_samples': [0.5,0.75,1.0], 
}

rf_grid= GridSearchCV(estimator= RandomForestRegressor(),
                     param_grid= param_grid,
                     cv=5,
                     verbose=2,
                     n_jobs=-1)


In [43]:
pipe= Pipeline(steps=[
    ('step1',trf),
    ('step2',StandardScaler()),
    ('step3',rf_grid)
])

# rf_grid.best_params_

In [44]:
pipe_rf= Pipeline(steps=[
    ('step1',trf),
    ('step2',StandardScaler()),
    ('step3',RandomForestRegressor( max_samples= 0.75, n_estimators= 100))
])

In [45]:
pipe_rf.fit(X_train,y_train)
y_predict= pipe_rf.predict(X_test)
print(r2_score(y_test,y_predict))
print(mean_absolute_error(y_test,y_predict))


0.9184042411315951
5.6515720400005565


### XGBRegressor

In [47]:
from xgboost import XGBRegressor
xgb= XGBRegressor(n_estimators=1000, learning_rate=0.1, max_depth=12)
pipe_xgb= Pipeline(steps=[
    ('step1',trf),
    ('step2',StandardScaler()),
    ('step3',xgb)
])

In [48]:
pipe_xgb.fit(X_train,y_train)
y_predict_xgb= pipe_xgb.predict(X_test)
print(r2_score(y_test,y_predict_xgb))
print(mean_absolute_error(y_test,y_predict_xgb))

0.948100716341297
3.9751518494973372


In [49]:
val_score_xgb=cross_val_score(pipe_xgb,X,y,cv=5)
print(val_score_xgb)
print(" ")
print("cross val_score mean XGB: " ,val_score_xgb.mean())

[0.94443314 0.94609528 0.94984039 0.94623632 0.94617197]
 
cross val_score mean XGB:  0.9465554186381115


### Here, we will take XGBRegressor as our ML model because it is having a better  r2_score( 0.948) and mean cross_val_score i.e. 0.946

### Testing models on new and recent data 

In [50]:
cols = ['Bat_first', 'Bat_second', 'Current_score', 'balls_left', 'wickets_left', 'crr', 'runs_last_five']
df_test = pd.DataFrame(columns=cols)

In [51]:
data = {
    'Bat_first': ['Ireland', 'Oman','Sri Lanka'],
    'Bat_second': ['India', 'Scotland','South Africa'],
    'Current_score': [73, 115, 68],
    'balls_left': [36,24 ,30],
    'wickets_left': [2, 5,3],
    'crr': [5.21, 7.18, 4.53],
    'runs_last_five': [29, 28, 28]
}
df_test=pd.DataFrame(data)

In [52]:
df_test

Bat_first    Bat_second  Current_score  balls_left  wickets_left   crr  \
0    Ireland         India             73          36             2  5.21   
1       Oman      Scotland            115          24             5  7.18   
2  Sri Lanka  South Africa             68          30             3  4.53   

   runs_last_five  
0              29  
1              28  
2              28

In [53]:
pipe_xgb.predict(df_test)    # 96, 150, 77

array([ 98.81416, 158.17311,  94.64323], dtype=float32)

In [54]:
pipe_rf.predict(df_test)

array([105.89, 154.7 ,  93.36])